modules/popup/react/lib/useCloseOnOutsideClick.ts

```diff
@@ -0,0 +1,32 @@
+import React from 'react';
+
+import {PopupStack} from '@workday/canvas-kit-popup-stack';
+
+/**
+ * Registers global lister for all clicks. It will only call the `onClose` callback if the click
+ * happened outside the `ref` element and its children _and_ the provided `ref` element is the topmost
+ * in the stack. The `ref` should be the same as the one passed to `usePopupStack` or the `Popper`
+ * component since `Popper` uses `usePopupStack` internally.
+ * @param ref
+ * @param onClose
+ */
+export const useCloseOnOutsideClick = <E extends HTMLElement>(
+  ref: React.RefObject<E>,
+  onClose: () => void
+) => {
+  const onClick = (event: MouseEvent) => {
+    if (
+      PopupStack.isTopmost(ref.current!) &&
+      !PopupStack.contains(ref.current!, event.target as HTMLElement)
```

Comment:
Workday's Standard Tooling calls this an "autohidePartner". It included the `owner` element and children as an "inside" click. This allows the application decide if clicking on a popup's target should toggle the existence of the popup or if it should stay open. Without this, clicking the target would toggle the popup and there would be no way to opt-out of the behavior. With this logic, it is an opt-in to toggle.

Reply:
This sounds like it's an important part of the codebase for future knowledge?

> Without this, clicking the target would toggle the popup and there would be no way to opt-out of the behavior. With this logic, it is an opt-in to toggle.

Reply:
I'll also add that to the JSDoc of `PopupStack.contains`

Reply:
<img width="649" alt="Screen Shot 2020-06-09 at 1 10 37 PM" src="https://user-images.githubusercontent.com/338257/84189536-a32e9280-aa52-11ea-9f59-9c4ccbf04fd7.png">